nni/algorithms/compression/pytorch/quantization/quantizers.py

```diff
@@ -426,74 +468,86 @@ def validate_config(self, model, config_list):
+            the real value to be quantized
+        scale : torch.Tensor
+            quantization scale
+        zero_point : torch.Tensor
+            quantization zero point
+        qmin : int
+            lower bound of the int range
+        qmax : int
+            upper bound of the int range
 
         Returns
         -------
         Tensor
         """
-        op.zero_point = op.zero_point.to(real_val.device)
-        op.scale = op.scale.to(real_val.device)
-        transformed_val = op.zero_point + real_val / op.scale
-        qmin = 0
-        qmax = (1 << bits) - 1
+        transformed_val = zero_point + real_value / scale
         clamped_val = torch.clamp(transformed_val, qmin, qmax)
         quantized_val = torch.round(clamped_val)
         return quantized_val
 
-    def _dequantize(self, op, quantized_val):
+    def _dequantize(self, quantized_val, scale, zero_point):
         """
         dequantize quantized value.
         Because we simulate quantization in training process, all the computations still happen as float point computations, which means we
         first quantize tensors then dequantize them. For more details, please refer to the paper.
 
         Parameters
         ----------
-        op : torch.nn.Module
-            target module
-        quantized_val : float
-            quantized_val value to be dequantized
+        quantized_val : torch.Tensor
+            the quantized value to be de-quantized
+        scale : torch.Tensor
+            quantization scale
+        zero_point : torch.Tensor
+            quantization zero point
 
         Returns
         -------
-        float
+        Tensor
         """
-        real_val = op.scale * (quantized_val - op.zero_point)
+        real_val = scale * (quantized_val - zero_point)
         return real_val
 
     def quantize_weight(self, wrapper, **kwargs):
-        config = wrapper.config
         module = wrapper.module
         weight = module.weight
-        weight_bits = get_bits_length(config, 'weight')
-        quant_start_step = config.get('quant_start_step', 0)
-        assert weight_bits >= 1, "quant bits length should be at least 1"
+        layer_quant_setting = module.layer_quant_setting
+        tensor_quant_setting = layer_quant_setting.weight
 
-        if quant_start_step > int(self.bound_model.steps):
-            return weight
+        # layer-wise settings
+        quant_start_step = layer_quant_setting.quant_start_step
+
+        # tensor-wise settings
+        dtype = tensor_quant_setting.quant_dtype
+        scheme = tensor_quant_setting.quant_scheme
+        qmin, qmax = tensor_quant_setting.get_qmin_qmax()
+        bits = tensor_quant_setting.bits
 
+        # In evaluation mode, we only quantize weight without updating statistics
         if not wrapper.training:
+            scale, zero_point = module.weight_scale, module.weight_zero_point
+            weight = self._quantize(weight, scale, zero_point, qmin, qmax)
```

Comment:
I am afraid that we must can only quantize the weight in each iteration. Because we can not in-place update the weight. So if we use ddp, the update will lost after each iteration.